pkg/virt-handler/vm.go

```diff
@@ -782,10 +786,46 @@ func (d *VirtualMachineController) heartBeat(interval time.Duration, stopCh chan
 				return
 			}
 			log.DefaultLogger().V(4).Infof("Heartbeat sent")
+			// Label the node if cpu manager is running on it
+			// This is a temporary workaround until k8s bug #66525 is resolved
+			featuregates.ParseFeatureGatesFromConfigMap()
+			if featuregates.CPUManagerEnabled() {
+				d.updateNodeCpuManagerLabel()
```

Comment:
Can you do both patch operations in one combined operation? We have to think about scale here too. For instance have a look at https://github.com/kubevirt/kubevirt/pull/1409/files#diff-9b18fe582fd0ecdc06b3fb28c3ad98b2R49. There, for different reasons I do something similar with constructiong a json-patch which conditionally adds or removes labels.

Reply:
Yea, I've seen it, but wasn't sure as I'm not familiar with it.
Another thing is that maybe we should turn features gate into a list instead of a string.

Reply:
regardless of the featuregate, it looks like updateNodeCpuManagerLabel is still only setting the label based on if the cpu-manager-policy=static in the kubelet command args.

I was under the impression we'd us the feature flag to automatically set the cpu manager label for now, and then transition to a dynamic detection of whether cpumanager is enabled per a node once we have a consistent way of doing so. Essentially we were forcing the user to both have a homogeneous cluster with respects to cpu manager and explicitly enable the feature gate.

By enabling the feature gate right now, users will be acknowledging that it is up to them to properly enable cpu manager on all nodes in order to use dedicated cpu placement. In the future when we eventually remove the feature gate, it will be because we can make this decision accurately for the user. 

is that reasonable for now?

Reply:
I don't think we can reasonably support the cpu manager auto detection logic today. There are already multiple ways to configure the kubelet outside of the cmdline args [1]. 

The feature gate is a way for us to remove KubeVirt from attempting to perform any sort of detection logic until k8s provides a supported interface for us to use in order to determine cpu manager is enabled. With the feature gate, we aren't taking responsibility for safety. That burden is on the user until we have a supportable way to auto detect cpu manager. The feature gate is saying, **"by enabling this, you understand that cpu manager must be enabled on all nodes in the cluster or bad things will happen"**.  Users willing to accept that condition have to explicitly enable the feature gate to use dedicated cpu placement. 

Once we have the supportable interface for determining cpu manager's policy, we can turn the feature gate on by default in a future kubevirt release and no longer require users to explicitly enable this support on our end. The feature gate is a stop gap for us until we have the k8s support we need to properly detect cpu manager on every node. 

1. https://kubernetes.io/blog/2018/07/11/dynamic-kubelet-configuration/

Reply:
I didn't see anywhere that enabling the cpu manager is supported via dynamic configuration.
All of the existing configuration show that on k8s[1] you have to update kubelet params manually and on Openshift[2] you must update the node-config-compute config map.
However, what this does in reality, is that each node runs a background script that reads the config map and updates kubelet params manually and restarts it.
According to this, we should find these parameters on the kubelet process.

[1] https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/
[2] https://docs.okd.io/latest/scaling_performance/using_cpu_manager.html

Reply:
I'm going to stop arguing. :)
If the consensus is to remove it, I'll remove it.
I'm not attached to much, it's just I think it's safer.

Reply:
> The feature gate is a way for us to remove KubeVirt from attempting to perform any sort of detection logic until k8s provides a supported interface for us to use in order to determine cpu manager is enabled. With the feature gate, we aren't taking responsibility for safety. That burden is on the user until we have a supportable way to auto detect cpu manager. The feature gate is saying, "by enabling this, you understand that cpu manager must be enabled on all nodes in the cluster or bad things will happen". Users willing to accept that condition have to explicitly enable the feature gate to use dedicated cpu placement.

@davidvossel I think I'm missing something. Thinking about this again. 
The feature gate is global to all nodes. 
According to your suggestion, by enabling it we will label all the nodes - even if there is no cpu manager running. There is no way to disable it for a certain host.
In that case, the user will have to run the cpu manager on all the nodes. This is not straightforward at all IMHO.
On the other hand, why do we need the feature gate at all in this case? We can simply let the users label the relevant nodes manually and not to force them to run the cpu manager on all hosts.

However, this is exactly what we have now. If the gate is enabled - we will do auto detection.
If the feature gate is disabled, don't detect but let the user label the nodes where he knows cpu manager is running.

Reply:
> the user will have to run the cpu manager on all the nodes. 

yep, that's the condition the user would have to accept when enabling the feature gate. The user would be committing to a homogeneous cluster with respects to cpu manager policy.  

> On the other hand, why do we need the feature gate at all in this case? We can simply let the users label the relevant nodes manually and not to force them to run the cpu manager on all hosts.

@vladikr I'd be fine with that. So in this case, we'd remove the feature gate and the autodetect logic. Users then have to set the label on nodes that they have cpu manager enabled on.

If we do this, then we can follow up on a discussion around whether or not to auto detect cpu manager afterwards. This lets us decouple the part we're in disagreement about from this PR so we can merge the pieces we're in agreement about. Essentially it gives us a supportable stop gap that can segue into more sophisticated detection logic later.